eclass/esed.eclass

```diff
@@ -0,0 +1,265 @@
+# arguments must be passed in the listed order with files last.  Each -e will
+# be a separate sed(1) call to evaluate changes of each.
+esed() {
+	(( ${#} >= 2 )) || die "too few arguments for ${_esed_cmd[0]:-${FUNCNAME[0]}}"
+
+	local endopts=false args=() contents=() exps=() files=()
+	local -i i
+	for ((i=1; i<=${#}; i++)); do
+		if [[ ${!i} =~ ^- ]] && ! ${endopts}; then
+			case ${!i} in
+				--) endopts=true ;;
+				-E|-n|-r) args+=( ${!i} ) ;;
+				-e)
+					i+=1
+					[[ ${!i} ]] || die "missing argument to -e"
+					exps+=( "${!i}" )
+					;;
+				*) die "unrecognized option for ${FUNCNAME[0]}" ;;
+			esac
+		elif (( ! ${#exps[@]} )); then
+			exps+=( "${!i}" ) # like sed, if no -e, first non-option is exp
+		else
+			[[ -f ${!i} ]] || die "not a file: ${!i}"
+			files+=( "${!i}" )
+			contents+=( "$(<"${!i}")" ) || die "failed reading: ${!i}"
+		fi
+	done
+	(( ${#files[@]} )) || die "no files in ${FUNCNAME[0]} arguments"
+
+	if [[ ${_esed_output} ]]; then
+		(( ${#files[@]} == 1 )) || die "${_esed_cmd[0]} needs exactly one input file"
+
+		# swap file for output to simplify sequential sed'ing
+		cp -- "${files[0]}" "${_esed_output}" || die
+		files[0]=${_esed_output}
+	fi
+
+	local changed exp newcontents sed
+	for exp in "${exps[@]}"; do
+		sed=( sed -i "${args[@]}" -e "${exp}" -- "${files[@]}" )
+		[[ ${ESED_VERBOSE} ]] && einfo "${sed[*]}"
+
+		"${sed[@]}" </dev/null || die "failed: ${sed[*]}"
+
+		changed=false
+		for ((i=0; i<${#files[@]}; i++)); do
+			newcontents=$(<"${files[i]}") || die "failed reading: ${files[i]}"
+
+			if [[ ${contents[i]} != "${newcontents}" ]]; then
+				changed=true
+
+				[[ ${ESED_VERBOSE} ]] || break
+
+				diff -u --color --label="${files[i]}"{,} \
+					<(echo "${contents[i]}") <(echo "${newcontents}")
+			fi
+		done
+
+		${changed} \
+			|| die "no-op: ${FUNCNAME[0]} ${*}${_esed_cmd[0]:+ (from: ${_esed_cmd[*]})}"
+	done
+}
+
+# @FUNCTION: enewsed
+# @USAGE: <esed-argument>... <output-file>
+# @DESCRIPTION:
+# esed() wrapper to save the result to <output-file>.  Intended to replace
+# ``sed ... input > output`` given esed() does not support stdin/out.
+enewsed() {
+	local _esed_cmd=( ${FUNCNAME[0]} "${@}" )
+	local _esed_output=${*: -1:1}
+	esed "${@:1:${#}-1}"
+}
+
+# @FUNCTION: erepl
+# @USAGE: <string> <replacement> <file>...
+# @DESCRIPTION:
+# Do basic bash-only ``${<file>//"<string>"/<replacement>}`` per-line
+# replacement in files(s).  Dies if no changes were made.  Suggested over
+# sed(1) where possible for simplicity and avoiding issues with delimiters.
+# Warning: erepl-based functions strip null bytes, use for text only.
+erepl() {
+	local _esed_cmd=( ${FUNCNAME[0]} "${@}" )
+	ereplp '.*' "${@}"
+}
+
+# @FUNCTION: enewrepl
+# @USAGE: <erepl-argument>... <output-file>
+# @DESCRIPTION:
+# erepl() wrapper to save the result to <output-file>.
+enewrepl() {
+	local _esed_cmd=( ${FUNCNAME[0]} "${@}" )
+	local _esed_output=${*: -1:1}
+	ereplp '.*' "${@:1:${#}-1}"
```

Comment:
What's this supposed to do? Write multiple files into one file? oO

Reply:
You mean about using `${#}-1` over fixed count to strip last argument (the output file). These wrappers are copy/pasted and esed has variable amount of arguments so felt simpler than trying to be specific with count in each (called functions also already verify if args are right).

Albeit this one would always be exactly `enewrepl string replace file.in "${T}"/file` for file.in->${T}/file

These exist because `sed s/// input > output` is not that uncommon, not that it's super essential but I'd use it like I did in nvidia's examples on dev ML.

Reply:
> Albeit this one would always be exactly `enewrepl string replace file.in "${T}"/file` for file.in->${T}/file

Exactly. So instead of trying to be lazy smartass about it, just put the arguments verbatim here and in the docstring for better readability.

Reply:
Would've been nice to say that in the first place so I wouldn't have to try guessing what you meant.

Reply:
And actually no, I don't want this. Passing too many arguments leads to checking for errors in the subsequent call without having to plaster args checks everywhere (other than erepld they don't even need special handling, it's all the same).

Reply:
Albeit I think some cases could be handled better, I'll have a look.

Reply:
Oh nevermind, it's already tested in tests/esed.sh for passing too many files to enewrepl and it dies fine (thought I forgot to check this for a sec). The "exactly one input" handles this